A couple of SourceRange fixes for `nonisolated`

hamishknight opened the pull request:
- Modifiers like `nonisolated` aren't spelled with an `@`, make sure the type attribute logic handles a missing `@`. I don't have a test case for this, but this would assert if the source range were queried.
- Make sure we correctly set the attribute range when used as a decl attribute with `@`.

rdar://147965036

hamishknight commented:
@swift-ci please smoke test

xedin commented:
@hamishknight I do have a test-case :)

```
func test(_: () -> Void) {}

test { @nonisolated in
}
```

This is a reduction from rdar://147965036

hamishknight commented:
> I do have a test-case :)

Funnily enough that's actually a different issue since it's a decl attribute there 😄 I see the issue, I can put up a fix for it

xedin commented:
Heh, thanks!

hamishknight committed:
[Parse] Set `AttrRange` for `@nonisolated`

Make sure we set the `AttrRange` in cases where `@` is present, but
`unsafe` isn't present.

rdar://147965036

hamishknight committed:
[AST] Handle missing `@` location in AtTypeAttrBase

Attributes like `nonisolated` aren't spelled with an `@`. I don't
have a test case, but this would assert if the source range were
queried.

hamishknight committed:
Fix spacing in diagnostic

Avoid printing a spurious space when not printing the `attribute`
branch of this diagnostic.

hamishknight commented:
Updated with the fix for rdar://147965036

hamishknight commented:
@swift-ci please smoke test
